src/runtime-rs/crates/runtimes/virt_container/src/container_manager/container.rs

```diff
@@ -305,10 +306,65 @@ impl Container {
 
     pub async fn stop_process(&self, container_process: &ContainerProcess) -> Result<()> {
         let mut inner = self.inner.write().await;
+
+        // ignore the stop when container status has already been stopped.
+        let status = inner.init_process.get_status().await;
+        if status == ProcessStatus::Stopped {
+            return Ok(());
+        }
+
         inner
-            .stop_process(container_process, true)
+            .stop_process(container_process)
+            .await
+            .context("stop process")?;
+
+        if container_process.process_type == ProcessType::Container {
+            // delete the container when container process is stopped
+            self.delete_container(container_process, true)
+                .await
+                .context("delete container")?;
+        }
+        Ok(())
+    }
+
+    async fn delete_container(
+        &self,
+        container_process: &ContainerProcess,
+        force: bool,
+    ) -> Result<()> {
+        // remove container
+        let remove_request = agent::RemoveContainerRequest {
+            container_id: container_process.container_id.container_id.to_string(),
+            ..Default::default()
+        };
+
+        self.agent
+            .remove_container(remove_request)
             .await
-            .context("stop process")
+            .context("remove container")
+            .or_else(|e| {
+                if force {
+                    warn!(
+                        self.logger,
+                        "stop container: agent remove container failed: {}", e
+                    );
+                    Ok(agent::Empty::new())
+                } else {
+                    Err(e)
+                }
+            })?;
+
+        let mut inner = self.inner.write().await;
+
+        // cleanup rootfs and volumes, and notify resource manager to keep states persistent
+        let cleaned_rootfs = inner.clean_rootfs().await?;
```

Comment:
We could consider doing this with the `resource manager`. As for block volume and block rootfs cleanup, we need to use the `device manager `to do this. Then, we need to pass the `device manager` as the parameter for the cleanup function. But in the current implementation, it's unable to fetch `device manager`. So, my suggestion is to pass `Vec<Arc<dyn Rootfs>>` and `Vec<Arc<dyn Volume>>` to `resource manager`.

Reply:
Update.
Now the cleanup operation is doing in `ResourceManager`, and also the unremoved rootfs and volumes will be returned to `Container`. WDYT?